internal/app/app.go

```diff
@@ -0,0 +1,169 @@
+	pathToServerFiles := filepath.Join(path, image_folder)
+	fileStorage, err := repository.New(pathToServerFiles, log)
+	if err != nil {
+		log.Error("Can't create file storage", logger.M{
+			"error": err,
+		})
+		return nil, fmt.Errorf("can't create file storage: %s", err)
+	}
+	fileService := storage.New(fileStorage, log)
+
+	// It creates an API router and handler with the file service
+	// and publisher, and registers the router to the handler.
+	api_router := api.New(fileService, publisher, log)
+	api_handler := handler.New(log)
+	api_handler.Register(api_router)
+
+	// It creates a compressor with the logger.
+	compressor := compressor.New(log)
+
+	// It creates a job, job's context, cancel function for the worker using the logger.
+	jobContext, jobCancelFunc := context.WithCancel(context.Background())
+	job := worker.New(
+		worker.WithClient(rabbitClient),
+		worker.WithFileRepository(fileStorage),
+		worker.WithCompressor(compressor),
+		worker.WithCancel(jobCancelFunc),
+		worker.WithContext(jobContext),
+		worker.WithLogger(log),
+	)
+
+	server := server.New(api_handler)
+
+	return &App{
+		srv: server,
+		job: job,
+		log: log,
+	}, nil
+}
+
+// Start method is responsible for starting the server and background job.
+func (a *App) Start() {
+	// Start the background job.
+	a.log.Info("Starting background job", nil)
+	a.job.Start()
+
+	// Launch the server in a separate goroutine.
+	a.log.Info("Starting server", logger.M{
+		"address": a.srv.Addr,
+	})
+	go func() {
+		if err := a.srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			a.log.Fatal("Server starting error.", logger.M{
+				"error": err,
+			})
+		}
+	}()
+}
+
+func (a *App) Stop() error {
+	// Stop background job
+	a.log.Info("Stopping background job", nil)
+	a.job.Stop()
+
+	// Close keep-alive connections
+	a.log.Info("Closing keep-alive connections", nil)
+	a.srv.SetKeepAlivesEnabled(false)
+
+	// Shutdown server with a timeout
+	a.log.Info("Shutdown server . . . Timeout", logger.M{
+		"timeout": timeoutDuration,
+	})
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutDuration)
+	defer cancel()
+
+	if err := a.srv.Shutdown(ctx); err != nil {
+		a.log.Error("Error shutdown", logger.M{
+			"error": err,
+		})
+		return err
+	}
+
+	a.log.Info("Server exiting", nil)
+
+	return nil
+}
+
+/*
+WaitForShutdown function sets up a channel to listen for signals
+indicating that the server should be shut down.
+
+It logs when it begins waiting for shutdown
+and when it receives a signal to terminate the server.
+*/
+func (a *App) WaitForShutdown() {
+	a.log.Info("Waiting for shutdown", nil)
+
+	/*
+		Specifically, it creates a channel to listen for SIGINT and SIGTERM signals,
+		which are sent to the process when the user requests to terminate the program
+		(e.g. by pressing Ctrl+C in the terminal).
+
+		Once the signal is received, the function logs the signal
+		that was received and the server will begin the shutdown process.
```

Comment:
How will it shut down the process if it's not notifying anyone?

https://www.rudderstack.com/blog/implementing-graceful-shutdown-in-go/